File: Heroku/test/src/dsl/PlanGroup.js
import Group from './Group';
import Plan from './Plan';

export default class PlanGroup extends Group {

	/**
	 * @param {ElementFinder} root
	 */
	constructor(root) {
		super(root, '.plans', '.plan');
		this.moreN = root.element(by.className('plans-footer')).element(by.partialButtonText("more ..."));
		this.moreAll = root.element(by.className('plans-footer')).element(by.partialButtonText("All"));
		this.scheduleExplorer = root.element(by.className('schedule-explorer'));
	}

	/**
	 * @returns {Plan}
	 */
	get(index) {
		return new Plan(this.items.get(index));
	}

	/**
	 * @param {function(Plan)} func
	 */
	each(func) {
		this.items.each((item) => func(new Plan(item)));
	}

	// TODO this doesn't work yet, expects after this don't see this.list
	listPlans() {
		this.scheduleExplorer.isSelected().then((isSelected) => {
			if (isSelected) {
				// selected means it's checked, so click to un-check
				this.scheduleExplorer.click();
			} else {
				// not selected, so it's already un-checked
			}
		});
	}
}

File: Heroku/test-integration/src/main/kotlin/net/twisterrob/cinema/frontend/test/pages/planner/PlanGroup.kt
package net.twisterrob.cinema.frontend.test.pages.planner

import com.paulhammant.ngwebdriver.ByAngular
import org.openqa.selenium.By
import org.openqa.selenium.WebElement

class PlanGroup(
	root: WebElement,
) : Group(root, ".plans", ".plan") {

	val footer: WebElement
		get() = root.findElement(By.className("plans-footer"))

	val moreN: WebElement
		get() = footer.findElement(ByAngular.partialButtonText("more ..."))

	val moreAll: WebElement
		get() = footer.findElement(ByAngular.partialButtonText("All"))

	val scheduleExplorer: WebElement
		get() = root.findElement(By.className("schedule-explorer"))

	val plans: List<Plan>
		get() = this.items.map(::Plan)

	operator fun get(index: Int): Plan =
		Plan(this.items[index])

	fun tryListPlans() {
		if (this.scheduleExplorer.isSelected) {
			// selected means it's checked, so click to un-check
			this.scheduleExplorer.click()
		} else {
			// not selected, so it's already un-checked
		}
	}
}
